corehq/apps/hqwebapp/utils.py

```diff
@@ -81,3 +86,70 @@ def aliased_language_name(lang_code):
+
+
+def extract_password(password):
+    # Passwords set with expected salts length and padding would respect this regex
+    reg_exp = r"^sha256\$([a-z|0-9|A-Z]{6})(.*)([a-z|0-9|A-Z]{6})=$"
+    match_result = re.match(reg_exp, password)
+    # strip out outer level padding of salts/keys and ensure three matches
+    if match_result and len(match_result.groups()) == 3:
+        match_groups = re.match(reg_exp, password).groups()
+        hash_left = match_groups[0]
+        hash_right = match_groups[2]
+        stripped_password = match_groups[1]
+        # decode the stripped password to get internal block
+        # decoded(salt1 + encoded_password + salt2)
+        try:
+            decoded_password = base64.b64decode(stripped_password)
+        except TypeError:
+            return ''
+        match_result_2 = re.match(reg_exp, decoded_password)
+        # strip out hashes from the internal block and ensure 3 matches
+        if match_result_2 and len(match_result_2.groups()) == 3:
+            match_groups = match_result_2.groups()
+            # ensure the same hashes were used in the internal block as the outer
+            if match_groups[0] == hash_left and match_groups[2] == hash_right:
+                # decode to get the real password
+                password_hash = re.match(reg_exp, decoded_password).groups()[1]
+                # return password decoded for UTF-8 support
+                try:
+                    return base64.b64decode(password_hash).decode('utf-8')
+                except TypeError:
+                    return ''
+            else:
+                # this sounds like someone tried to hash something but failed so ignore the password submitted
+                # completely
+                return ''
+        else:
+            # this sounds like someone tried to hash something but failed so ignore the password submitted
+            # completely
+            return ''
+    else:
+        # return the password received AS-IS
+        return password
+
+
+# Memoized for multiple decode attempts in the same request:
+# 1. an attempt to decode a password should be done just once in a request for the login attempt
+# check to work correctly.
+# 2. there should be no need to decode a password multiple times in
+# the same request.
+@memoized
+def decode_password(password, username=None):
+    if settings.ENABLE_PASSWORD_HASHING:
+        if username:
+            # To avoid replay attack where the same hash used for login is used on attack
+            if HashedPasswordLoginAttempt.objects.filter(
+                username=username,
+                password_hash=password,
+                used_at__gte=(datetime.today() - timedelta(HASHED_PASSWORD_EXPIRY))
+            ).exists():
+                return ''
+            else:
+                HashedPasswordLoginAttempt.objects.create(
+                    username=username,
+                    password_hash=password
```

Comment:
Yikes! That thought was in my mind when i was doing it, not sure why i didn't connect it to what i was implementing(may be coz it was encoded with some algorithm). Thanks for the correction. Bad miss!